base/compiler/optimize.jl

```diff
@@ -317,12 +312,6 @@ function stmt_effect_flags(𝕃ₒ::AbstractLattice, @nospecialize(stmt), @nospe
             f = argextype(args[1], src)
             f = singleton_type(f)
             f === nothing && return (false, false, false)
-            if f === UnionAll
-                # TODO: This is a weird special case - should be determined in inference
-                argtypes = Any[argextype(args[arg], src) for arg in 2:length(args)]
-                nothrow = _builtin_nothrow(𝕃ₒ, f, argtypes, rt)
-                return (true, nothrow, nothrow)
-            end
```

Comment:
`UnionAll` isn't a builtin function, so its `:nothrow` modeling
shouldn't live in `_builitin_nothrow`. The existence of its special
casing within `stmt_effect_flags` is also weird, and I confirmed it
doesn't actually give us any refinement withe following patch, so is
safe to be removed:
```diff
diff --git a/base/compiler/optimize.jl b/base/compiler/optimize.jl
index 85942d3ca8..9e17832d4f 100644
--- a/base/compiler/optimize.jl
+++ b/base/compiler/optimize.jl
@@ -294,7 +294,9 @@ end

 Returns a tuple of `(:consistent, :removable, :nothrow)` flags for a given statement.
 """
-function stmt_effect_flags(𝕃ₒ::AbstractLattice, @nospecialize(stmt), @nospecialize(rt), src::Union{IRCode,IncrementalCompact})
+function stmt_effect_flags(𝕃ₒ::AbstractLattice, @nospecialize(stmt), @nospecialize(rt),
+                           src::Union{IRCode,IncrementalCompact},
+                           original_flag::Union{Nothing,UInt32}=nothing)
     # TODO: We're duplicating analysis from inference here.
     isa(stmt, PiNode) && return (true, true, true)
     isa(stmt, PhiNode) && return (true, true, true)
@@ -318,9 +320,19 @@ function stmt_effect_flags(𝕃ₒ::AbstractLattice, @nospecialize(stmt), @Nospe
             f = singleton_type(f)
             f === nothing && return (false, false, false)
             if f === UnionAll
-                # TODO: This is a weird special case - should be determined in inference
-                argtypes = Any[argextype(args[arg], src) for arg in 2:length(args)]
-                nothrow = _builtin_nothrow(𝕃ₒ, f, argtypes, rt)
+                nothrow = false
+                if length(args) == 3
+                    argtype3 = argextype(args[3], src)
+                    if !isvarargtype(argtype3)
+                        argtype2 = argextype(args[2], src)
+                        if argtype2 ⊑ TypeVar && argtype3 ⊑ Type
+                            nothrow = true
+                            https://github.com/Assert (original_flag !== nothing &&
+                                has_flag(original_flag, IR_FLAG_NOTHROW) &&
+                                has_flag(original_flag, IR_FLAG_CONSISTENT)) "refinement detected"
+                        end
+                    end
+                end
                 return (true, nothrow, nothrow)
             end
             if f === Intrinsics.cglobal || f === Intrinsics.llvmcall
@@ -377,9 +389,10 @@ function stmt_effect_flags(𝕃ₒ::AbstractLattice, @nospecialize(stmt), @Nospe
 end

 function recompute_effects_flags(𝕃ₒ::AbstractLattice, @nospecialize(stmt), @nospecialize(rt),
-                                 src::Union{IRCode,IncrementalCompact})
+                                 src::Union{IRCode,IncrementalCompact},
+                                 original_flag::Union{Nothing,UInt32}=nothing)
     flag = IR_FLAG_NULL
-    (consistent, removable, nothrow) = stmt_effect_flags(𝕃ₒ, stmt, rt, src)
+    (consistent, removable, nothrow) = stmt_effect_flags(𝕃ₒ, stmt, rt, src, original_flag)
     if consistent
         flag |= IR_FLAG_CONSISTENT
     end
diff --git a/base/compiler/ssair/inlining.jl b/base/compiler/ssair/inlining.jl
index 28b79cbfd3..db41b76bbb 100644
--- a/base/compiler/ssair/inlining.jl
+++ b/base/compiler/ssair/inlining.jl
@@ -1218,7 +1218,7 @@ end
 add_inst_flag!(inst::Instruction, ir::IRCode, state::InliningState) =
     add_inst_flag!(inst, ir, optimizer_lattice(state.interp))
 function add_inst_flag!(inst::Instruction, ir::IRCode, 𝕃ₒ::AbstractLattice)
-    flags = recompute_effects_flags(𝕃ₒ, inst[:stmt], inst[:type], ir)
+    flags = recompute_effects_flags(𝕃ₒ, inst[:stmt], inst[:type], ir, inst[:flag])
     add_flag!(inst, flags)
     return !iszero(flags & IR_FLAGS_REMOVABLE)
 end
diff --git a/base/compiler/tfuncs.jl b/base/compiler/tfuncs.jl
index 92b03964a1..3b3588a7da 100644
--- a/base/compiler/tfuncs.jl
+++ b/base/compiler/tfuncs.jl
@@ -2177,7 +2177,7 @@ end
 end

 # Query whether the given builtin is guaranteed not to throw given the argtypes
-@nospecs function _builtin_nothrow(𝕃::AbstractLattice, f, argtypes::Vector{Any}, rt)
+function _builtin_nothrow(𝕃::AbstractLattice, @nospecialize(f::Builtin), argtypes::Vector{Any}, @nospecialize(rt))
     ⊑ = partialorder(𝕃)
     if f === memoryref
         return memoryref_builtin_common_nothrow(argtypes)
@@ -2225,8 +2225,6 @@ end
     elseif f === (<:)
         na == 2 || return false
         return subtype_nothrow(𝕃, argtypes[1], argtypes[2])
-    elseif f === UnionAll
-        return na == 2 && (argtypes[1] ⊑ TypeVar && argtypes[2] ⊑ Type)
     elseif f === isdefined
         return isdefined_nothrow(𝕃, argtypes)
     elseif f === Core.sizeof
```